x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java

```diff
@@ -0,0 +1,260 @@
+        if (signature != null) {
+            if (sigAlg == null) {
+                logAndThrow(() ->
+                    new ParameterizedMessage("Query string [{}] contains a Signature but SigAlg parameter is missing",
+                        queryString));
+            }
+            final X509Credential spSigningCredential = sp.getSigningCredential();
+            if (spSigningCredential == null) {
+                logAndThrow(
+                    "Unable to validate signature of authentication request, Service Provider hasn't registered signing credentials");
+            }
+            validateSignature(samlRequest, sigAlg, signature, sp.getSigningCredential(), relayState);
+        }
+        validateAuthnRequest(authnRequest, sp);
+        Map<String, Object> additionalData = buildAdditionalData(authnRequest, sp);
+        return new SamlValidateAuthnRequestResponse(sp.getEntityId(), authnRequest.isForceAuthn(), additionalData);
+    }
+
+    private Map<String, Object> buildAdditionalData(AuthnRequest request, SamlServiceProvider sp) {
+        Map<String, Object> data = new HashMap<>();
+        final NameIDPolicy nameIDPolicy = request.getNameIDPolicy();
+        if (null != nameIDPolicy) {
+            final String requestedFormat = request.getNameIDPolicy().getFormat();
+            if (Strings.hasText(requestedFormat)) {
+                data.put("nameid_format", requestedFormat);
+                // we should not throw an error. Pass this as additional data so that the /saml/init API can
+                // return a SAML response with the appropriate status (3.4.1.1 in the core spec)
+                if (requestedFormat.equals(UNSPECIFIED) == false &&
+                    requestedFormat.equals(sp.getNameIDPolicyFormat()) == false) {
+                    logger.warn(() ->
+                        new ParameterizedMessage("The requested NameID format [{}] doesn't match the allowed NameID format" +
+                            "for this Service Provider is [{}]", requestedFormat, sp.getNameIDPolicyFormat()));
+                    data.put("error", "invalid_nameid_policy");
+                }
+            }
+        }
+        return data;
+    }
+
+    private void validateAuthnRequest(AuthnRequest authnRequest, SamlServiceProvider sp) {
+        checkDestination(authnRequest);
+        checkAcs(authnRequest, sp);
+    }
+
+    private void validateSignature(String samlRequest, String sigAlg, String signature, X509Credential credential,
+                                   @Nullable String relayState) {
+        try {
+            final String queryParam = relayState == null ?
+                "SAMLRequest=" + urlEncode(samlRequest) + "&SigAlg=" + urlEncode(sigAlg) :
+                "SAMLRequest=" + urlEncode(samlRequest) + "&RelayState=" + urlEncode(relayState) + "&SigAlg=" + urlEncode(sigAlg);
+            Signature sig = Signature.getInstance("SHA256withRSA");
```

Comment:
Shouldn't this reference `sigAlg`? or at least, should we validate & require that `sigAlg` is SHA256+RSA? (Ir we only want to support 1 algorithm).

Reply:
:man_facepalming: